### ISSUE TITLE
Missunderstandings in G-Code interpretation

### ISSUE THREAD

**Issue by VirToReal** [MarlinFirmware/Marlin#1954]
Hello everybody,

i made an CBC-"Milling" Machine by myself. It also can print in 3D if i had made an extruder ;-). I decided to use Marlin as Firmware on an Arduino-Board with a RAMPS-Shield 1.4 on it. It works pretty well but i got some annoying fails on milling with Marlin (i already know: "its not designed for that")...

At First:
I used some Python-Script to convert PyCam ngc Code into RepRap-Gcode for 3D Printers (scaled-down settings / add Z axis and movement Speeds while Cutting / Traveling). The Result could be watched here: https://www.dropbox.com/s/iyifm093criudfr/VirToGravTest4_converted.gcode?dl=0 (converted)
Original: https://www.dropbox.com/s/em8l3p4eqd9uimg/VirToGravTest4.ngc?dl=0

It works pritty awesome except of these 4 faults:
![dscn3120](https://cloud.githubusercontent.com/assets/10797066/7261067/26cb6880-e86f-11e4-9c60-3ba8eb6b5ef2.JPG) (Dont mention that letters in the background)

Vir(T)oGrav:
The letter "T" got an smaller right "outrigger"

VirT(o)Grav:
The surrounding of the letter "O" dont stop at its starting point

VirToG(r)av:
The surrounding of the letter "r" dront stop at its starting point

VirToGr(a)v:
The "a" also got some awkward issues with its surrounding 

I dont know why this happens, but it happens all the time, i made this milling 18x Times with always the same Results. I tried different speeds, differed jerk values, different highs, at last i let a pen draw the lines ;-).. always the same result.

Then i got a clou from a friend of mine:
I use auto-bed-leveling to recalculate the large milling area (600x380mm) because its terrible uneven. With auto-bed-leveling its quiet great! (awesome feature btw.). The Idea was, these calculation in X Y because of the uneven bed will cause fluctuations in calculating the axes (because auto-bed-leveling is still under process (i thought)). So i tried without -> same result ;-(

I'm out of ideas and tired of this Problem... the gcode looks okay to me, cant see the issue until it became reality. Maybe someone can see something, maybe its an issue in Marlin.. 
I can only guarantee one thing: it is not of mechanical nature

**Comment by brainscan** [comment]
I think it's the gcode that's the problem, there are some interesting moves in it. Have you only tried this one file because if so that would make the most sense. You are using marlin in an unconventional way so I don't think it's really an issue with the firmware. Most of the part you have pictured looks good so I'd assume marlin is working correctly and is trying to interpret bad code. That happens when slicing parts for printing too so sometimes you have to make a new file or correct the code manually. I don't have any way to test your code at the moment to see if that's the problem, sorry.

**Comment by tonnico** [comment]
gcode looks ok for me. You should tighten your belt. This is maybe to loose and you have a backlash.

**Comment by brainscan** [comment]
I've never seen a loose belt give exactly the same result 18x in a row at different speed settings. I have seen bad code run over and over again with the user expecting different results, which is also a sign of madness.

**Comment by tonnico** [comment]
Yes, right. Just took a look at the gcode. If that would be the issue, the end at the VirT(o)Grav should be under not over the starting point.

What are your steps/mm? What kind of driver you are using?

**Comment by brainscan** [comment]
Have you tried running the code through a simulator/visualiser? That would confirm or not if its your problem, when I get home later I might copy and paste your code into something like slic3r or repetier and see if likes it or not. There are plenty of free ways to check it so if they don't work I'll try something else. 

An alternative might be to create the file using slicing software and then just manually alter the code so it mills down rather than printing up, hope that makes sense.

**Comment by tonnico** [comment]
I just visualize the code. The code isn't the problem.
My guess it's something with the slow speed.

**Comment by brainscan** [comment]
I find that odd that after 18 attempts with various speed settings you think it's a speed issue, especially as apparently all results are the same. In my experience if it's a belt or speed issue the results are never exactly the same, sorry if that's not what you have experienced.

**Comment by tonnico** [comment]
It depends on his setup. How much steps he hast for each mm. At such slow speeds Marlin can have issues. The interrupt works only on 16bit. At 2MHz it means the steprate is limited to min 32steps/s. His max speed could be at ~250 steps/s. This could makes the problem. It's just an idea...

**Comment by VirToReal** [comment]
Thanks for the interest of my problem!

I visualized the Gcode in different free programs and cant find any problem in the G-Code (i should had mentioned that, sorry ;-). Brainscan is right about belt problems, i also adjusted the tautness with no changes on the end-result.

But your comment Wurstnase about drivers, i got different types on the X any Y Axis. Maybe this can be a great explanation for my worrys? The Y and Z are driven by a DRV8825 (2x Y Axis on 1xStepperengine / 1x Z Axis 2x Stepperengine). The X Axis is driven by an China-Shit driver which came with the arduino. I used that because of the low current i needed for that "easy" Axis...
![2015-02-10 20 42 38](https://cloud.githubusercontent.com/assets/10797066/7279508/c86ec436-e91c-11e4-8fd9-a029103b6c77.jpg)
 i think that was an misstake :-(... i was convinced every stepper counts the same distance. I'm not sure about that now ;-)

Stepping:
X: 80 Y: 80 Z: 160 steps/mm
Max Feedrate:
X: 250 Y: 250 Z: 4
Microstepping:
X: 1/16 Y: 1/16 Z: Full-Step

I made best results on milling with 150mm/min -> 0,5mm Graver,  so the engraving on the picture mentioned above. 

Would it help to change the X-Axis Driver to the same one like the others?

**Comment by nophead** [comment]
For some reason your pictures just show an error message.

Different drivers shouldn't matter as long as they work and don't skip steps.

**Comment by brainscan** [comment]
Hmm, yeah I hadn't really thought about that. I also hadn't really thought about the extra force needed when milling and the fact it's using belts would show this sort of thing if they were stretching. I'm sorry for thinking it was the code too, I had a quick look on an online viewer and couldn't see a problem. Just seemed most likely to me but I was mistaken. 

Strange that it's so consistent and that some letters look great and others not so.

**Comment by VirToReal** [comment]
I will try this out with a pen and faster "print" speed. 

Brainscan: i also got this problem with a Pen, dont think the belt cause this problem. But i tried only at 150mm/min. I will scale this up and report, you will hear from me tomorrow after work ;-)

**Comment by brainscan** [comment]
I know from your previous post that you had used a pen with the same result so I was fairly sure you had them tight, also because you had tried different speeds I thought that wasn't the problem. I hadn't thought that if all the speeds are still below a certain threshold you'd get the same error. I think wurstnase is on to something as I'd forgotten about how weird my printer goes if I accidentally ask it to go really slowly. Sorry for basically being unhelpful but I am interested in what's causing it.

**Comment by VirToReal** [comment]
I made a quick test with the following result:
![img_20150423_184618](https://cloud.githubusercontent.com/assets/10797066/7305579/43b87c62-e9ff-11e4-8ea7-6ea50333d03b.jpg)
at the last test the x-axis china-driver quitted his duty... but this is not the problem, the Y axis drive various distances.. :( Same Problem as alway, at the same points.. the 150mm/min one just looks ugly because of the ink and its time to flow out of the pen

I will test tomorrow with disabled auto-bed-leveling in firmware

**Comment by brainscan** [comment]
Even though the x driver gave up it does seem like there are fewer errors (compared to the original picture) as the speed goes up, although it is still very slow. If you are only using a pen I would think 20-30mm/s would still be slow but shouldn't lead to the problem wurstnase mentioned. If that still isn't the problem then will have to try something else.

**Comment by VirToReal** [comment]
I replaced the defective driver with another Chinese driver and ran the test again at different speed written next to the "painting" in mm/min.
I also adjusted the jerk values during testing as noted on the sheet. The last one at the bottom is "printed" with auto-bed-leveling disabled in firmware. 

Btw.: these "prints" are made with mounted Dremel in place. So don't mention the "letter shivering." At the speed of 1500mm/min i stabilized the dremel on its wire during printing, because it's only mounted on its thread. Further test with speed improvements were against Max speed of the Z Axis, (because I just adjusted Speed Multiplier).

I still has no clue of that problem... 

![2015-04-24 22 16 38](https://cloud.githubusercontent.com/assets/10797066/7328992/73370d4c-eada-11e4-94a5-219fd1855545.jpg)

**Comment by thinkyhead** [comment]
We just increased the microstep delay to 2µs (for the sake of DRV8825 drivers). Hopefully this has no bearing on performance. As I understand it, using 32x microstepping will generate more activity than 16x microstepping and put more strain on the board. But you should still be within tolerances using a Mega2560.

**Comment by tonnico** [comment]
We don't have any delay at normal moving.

**Comment by thinkyhead** [comment]
Oh right, only in babystepping. I wonder if a delay would help, or is it already "delaying" in some other way, such as having some code instructions in between…?

**Comment by tonnico** [comment]
There is no real delay. At this slow speed, below 10.000 steps/s you can try the toshiba-mode.

**Comment by VirToReal** [comment]
I'm not using 1/32 microstepping mode, just 1/16 on the X Y Axis. The speed i noticed next to the pen-drawings on my last picture are in mm/min. The last one with 1500mm/min was the fastest one with the same issue. These would be 2000 steps/s. This is fast i think but not fast enough? What is that toshiba-mode?

**Comment by tonnico** [comment]
2000 step/s is ok.
https://github.com/MarlinFirmware/Marlin/blob/Development/Marlin/Configuration.h#L322

**Comment by VirToReal** [comment]
Hello again,

sorry it took some time, i updated to the newest firmware because the Toshiba-stepper Firmware seems to drive only one of my two Y-Axis Steppers. So i updated to the newest and tried again, same result. Just one Stepper (the original one on the Ramps 1.4 board) burst forward, the other one won't do a thing until i commit toshiba-drivers out again. 

Of course i tried with the newest firmware ;-) same result, just the sound on jerking sounds smoother ;-)

**Comment by tonnico** [comment]
Go in the stepper.cpp. At line ~580 you find something like this

```
           #ifdef CONFIG_STEPPERS_TOSHIBA
        /**
         * The Toshiba stepper controller require much longer pulses.
         * So we 'stage' decompose the pulses between high and low
         * instead of doing each in turn. The extra tests add enough
         * lag to allow it work with without needing NOPs
         */
        #define STEP_ADD(axis, AXIS) \
         _COUNTER(axis) += current_block->steps[_AXIS(AXIS)]; \
         if (_COUNTER(axis) > 0) { _APPLY_STEP(AXIS)(!_INVERT_STEP_PIN(AXIS),0); }
        STEP_ADD(x,X);
        STEP_ADD(y,Y);
        STEP_ADD(z,Z);
        #ifndef ADVANCE
          STEP_ADD(e,E);
        #endif

        #define STEP_IF_COUNTER(axis, AXIS) \
          if (_COUNTER(axis) > 0) { \
            _COUNTER(axis) -= current_block->step_event_count; \
            count_position[_AXIS(AXIS)] += count_direction[_AXIS(AXIS)]; \
            _APPLY_STEP(AXIS)(_INVERT_STEP_PIN(AXIS),0); \
          }

        STEP_IF_COUNTER(x, X);
        STEP_IF_COUNTER(y, Y);
        STEP_IF_COUNTER(z, Z);
        #ifndef ADVANCE
          STEP_IF_COUNTER(e, E);
        #endif

      #else // !CONFIG_STEPPERS_TOSHIBA
```

use this ^^ code and replace yours and test again with #define toshiba-stuff

**Comment by thinkyhead** [comment]
@Wurstnase Do you also recommend this as a permanent patch for Marlin, or just for certain hardware?

**Comment by tonnico** [comment]
I think there could be a better way which will work for Toshiba and other drivers without that #define.

The idea i have to put the 'Calculate new timer value' in the loop. Write step high, then if last step, calculate the new timer value, write step low. This will give a lot of lag in single stepping mode.

**Comment by tonnico** [comment]
@thinkyhead https://github.com/Wurstnase/Marlin/tree/new_double_stepping this is the idea i've talked.

**Comment by thinkyhead** [comment]
@Wurstnase Those changes seem comprehensive enough. I was wondering, with the loop gone, maybe this…

```
for(int8_t i=0; i < step_loops; i++) {
  advance -= advance_rate;
}
```

…can be replaced with…

```
if (step_loops > 0) advance -= advance_rate * step_loops;
```

**Comment by tonnico** [comment]
This needs also some work for the last step (break). I will prepare a PR soon.

**Comment by VirToReal** [comment]
Thanks Wurstnase, your patch worked well, but the result is the same. No differences noticeable:

![2015-05-01 16 05 41](https://cloud.githubusercontent.com/assets/10797066/7432673/61f6f100-f02a-11e4-889a-05716b1b0346.jpg)
Old an new Firmware "printed" also with a feedrate of 600mm/min

I noticed some differences (not related to toshiba-setting) between 2000steps/sek and 800steps/sek 

(left 800 / right 2000)
![t-compare](https://cloud.githubusercontent.com/assets/10797066/7432732/dbea551a-f02a-11e4-839b-57abc635ee29.jpg)
dont know why or whats the reason this always happens at low speeds at the same position

I also clipped out the O from VirT(o)Grav and align it at same distance and let pyCam gcode it again. The result looks fancier than bevor:
![2015-05-01 16 05 58](https://cloud.githubusercontent.com/assets/10797066/7432785/7b5548ee-f02b-11e4-92f0-c4d6f8e08a51.jpg) Again i dont know whats the difference between placed in the Word VirToGrav and Stand-Alone... why is there a different Result? 
 There are noticeable differences between the direction the circles are printed, i made a drawing on the left to hi-light that.

Again i made some tests as close as possible to the X / Y Engine (the others are as far away as possible) with the same result:
![2015-05-01 16 52 38](https://cloud.githubusercontent.com/assets/10797066/7432864/5019a692-f02c-11e4-9f4b-f9e5a7aee31a.jpg) i took the pen just a bit to far to the sheet ;-)

So, why is that? Got no ideas left :-(

**Comment by thinkyhead** [comment]
@HardRainbow Are you able to replace the pen with a ballpoint? It would make thinner lines, and it may be easier to see the movements.

**Comment by VirToReal** [comment]
here we go:
![dscn3122](https://cloud.githubusercontent.com/assets/10797066/7442623/c383e77c-f118-11e4-8b8c-b7422a4d5fa6.JPG)
nice to see here, how the letter T change his form like mentioned above between 400mm/min and 1200mm/min 

![dscn3123](https://cloud.githubusercontent.com/assets/10797066/7442636/6e32deda-f119-11e4-9441-77b4fc7c918c.JPG)

![dscn3124](https://cloud.githubusercontent.com/assets/10797066/7442637/779d1a08-f119-11e4-8ef1-fff0595f2427.JPG)

hope this will help in any way

**Comment by AnHardt** [comment]
What i see, looks like backlash.
Maybe it's not the belts, but tilling the z-axis under load.

**Comment by thinkyhead** [comment]
It is sometimes hard to isolate what might be mechanical issues, but begin by tightening everything – _except the belts_ (or other drive mechanisms). You want some tension on the belts, but not too much. If you have a way to reliably tension them that is one place to start. Of course, ensure there's no play in the movement axes. Check that when the Z level goes up, there's nothing lagging... all the usual tightening and isolation. If that doesn't help then start playing with acceleration, jerk, etc... within each speed range, and hopefully some consistent variable will start to jump out.

**Comment by VirToReal** [comment]
@AnHardt 
I already checked for backlash of any backlash of the Z Axis with an USB Microscope, there is just a littl settldown at one Side, i made a table of that:
![table](https://cloud.githubusercontent.com/assets/10797066/7482097/35191304-f375-11e4-8983-45fab49ed945.png)
its an older Table, cant find a newer one with nicer results jet ;-)

Measure of Z level before and after printing without any differences -> should be ok :)

@thinkyhead 
Youre right, it is very hard the find mechanical issues, but there are some things which let me guess this cant be of mechanical nature, the the start surrounding in letter VirToG(r)af wont fit with the end, but at the first r-letter Vi(r)ToGrav it works. No matter of speed nor position..

i tried something: i turned the SVG in Inkscape at 90° and run pyCam again, 
![dscn3127](https://cloud.githubusercontent.com/assets/10797066/7482239/4d78ab16-f376-11e4-976a-81fc8d694cf7.JPG)

i also increased the z-level on drawing on a hard drive ;-)
![dscn3128](https://cloud.githubusercontent.com/assets/10797066/7482278/8ada5856-f376-11e4-8332-2563a833042f.JPG)

same results :-(

@Wurstnase 
in toshiba-stepper mode (with your patch), after a "print", i manually change to position of the "extruder" with relative positioning -> 100mm -> he increases speed and at 2/3 of the way he slow down and looses steps after slowing down (on X and Y Axes). After reset of the arduino everything works fine again. I dont know if the reason is because of  toshiba-driver-setting or your patch, because i cant use this setting without your patch ;-) hope this description will help you :-S

At the End:
i ordered a new stepper for the x axis, so every driver would be of the same kind (DRV). I hope this will help (i dont guess so), you will hear from me soon ;-)

Greetings and thanks to everybody for the interest in my problem

**Comment by thinkyhead** [comment]
@HardRainbow Excellent to have it isolated to the software. This is the kind of _very subtle_ behavior that would be impossible to see in plastic, but which a ball-point makes so much clearer! The "T" shape is a good controlled test, so let's keep focusing on that little anomaly and see if we can isolate some cause. It might be some unavoidable artifact of the interaction between the mathematics, stepper timing, and mechanics, but fixable or unfixable, that's what we want to know.

Do the artifacts scale up and down also? Or, do they only appear when printing certain shapes at certain sizes? And how many segments are in those regions where the glitches appear? Are there a lot of very small segments, or is it a simple hard corner…?

**Comment by VirToReal** [comment]
Scale Up and Down Test:

Speed: 600mm/min
![dscn3129](https://cloud.githubusercontent.com/assets/10797066/7553353/0c652e52-f6f4-11e4-81b3-195ba9dc409a.JPG)
![dscn3130](https://cloud.githubusercontent.com/assets/10797066/7553354/12a78e9a-f6f4-11e4-8267-95e78f307b0a.JPG)
![dscn3131](https://cloud.githubusercontent.com/assets/10797066/7553355/168b13ba-f6f4-11e4-8720-54d7a4698e7d.JPG)

Speed: 1500mm/min (Detached Dremel)
![dscn3132](https://cloud.githubusercontent.com/assets/10797066/7553359/31aaf26e-f6f4-11e4-830a-1ef0fdf8facb.JPG)
![dscn3133](https://cloud.githubusercontent.com/assets/10797066/7553361/385dc49c-f6f4-11e4-9f77-deb59bd47b28.JPG)
![dscn3134](https://cloud.githubusercontent.com/assets/10797066/7553362/3d84a1a2-f6f4-11e4-8c9f-4aa4da615103.JPG)

I reused this logo and made up some new Shape-Sizes. The original one measures a size of 90mm width, so i made examples with 60 / 75 / 105 and 120mm (Resizing with Inkskape and slicing with pycam). At the size of 75 i forgot the "infill" ;-), i thought it wouldnt be necessary 

Every "print" in their speed category made in its original place, i just repositioned the sheet. The sheet at the speed of 1500mm/min is made at a different position. The results are exactly the same!!
I just noticed differences between 60 and 75:
![differences](https://cloud.githubusercontent.com/assets/10797066/7553447/c172ced0-f6f9-11e4-86a1-a48ada00d190.jpg)

Of course the GCodes are different (With different start and ending points):
https://www.dropbox.com/s/1i0z4dig9fezhvv/VirToGravTest4_resize_60_converted.gcode?dl=0
https://www.dropbox.com/s/mva4wqtl3qrgtna/VirToGravTest4_resize_75_converted.gcode?dl=0
https://www.dropbox.com/s/iay4l6akqeb6niq/VirToGravTest4_resize_105_converted.gcode?dl=0
https://www.dropbox.com/s/9e7pc0rxkdojtx9/VirToGravTest4_resize_120_converted.gcode?dl=0

I also made a Movie of the "engraving" of the 120mm one:
https://www.dropbox.com/s/t55mn8kfz3z0k5g/VID_20150508_175440.3gp?dl=0
Dont mention the "stucking" of the Z Axis, its the error i at fast speeds i already mentioned Wurstnase in a previous post because of the Toshiba-Setting and/or the replaced C-Code. I didnt set back to original yet. You can watch its couse at beginning and right bevor the letter "O"
Also dont mention the clicking, its the sound of the auto-bed-leveling because of the Z-Axis threads are guided by bearings at the top.  

these would be the resize test. Im looking forward for getting more ideas in mind to get closer to this problem. 

Collected Information yet:
Shapes changed its behaviour between the size of 60 / 75. 75/90/105/120 looks exactly the same. No changes between the speed of 600 / 1500 mm/min.

**Comment by nophead** [comment]
I think you simply have backlash. What type of pulleys are you using?

Note pulley backlash is not like gear backlash. You don't get a dead spot when you change direction because the teeth don't slip. What happens is when you change direction the new teeth engage at the opposite side of the valley. After half a turn in the new direction all the teeth are engaged on the new side and the distance travelled is a little bit short. So where it stops depends on which direction it came from and also how far it travelled in that direction. I.e. the backlash is distributed over a full half turn rather than happening immediately on direction change.

To check for it make a simple pattern such as a large H ( bigger than half a pulley revolution). You should see the verticals misaligned top to bottom.

**Comment by VirToReal** [comment]
@nophead, Good Idea, never thought about that!
I got GT2 Belts on a 20th Teeth Pulley -> 40mm each Turn 
i made some pattern for controlling if there is any backlash because a couldnt see any difference with the Letter H in different sizes:
![img_20150510_175213](https://cloud.githubusercontent.com/assets/10797066/7555090/08ba348a-f741-11e4-8422-c81e3a426dbd.jpg)
![img_20150510_175223](https://cloud.githubusercontent.com/assets/10797066/7555091/0cf14386-f741-11e4-83ee-dd35dbc4b715.jpg)
This one was "printed" 6xTimes:
![img_20150510_180031](https://cloud.githubusercontent.com/assets/10797066/7555096/1de37506-f741-11e4-8ff6-693e2ba6f27b.jpg)

G-Code File with randomly printing of Squares:
https://www.dropbox.com/s/7oyn0kkge5cskdb/PatternV2_converted.gcode?dl=0

Testprint with X-Boxes:
![img_20150510_192732](https://cloud.githubusercontent.com/assets/10797066/7555437/be2351f8-f74b-11e4-8b46-307857025d65.jpg)

G-Code File with randomly printing of X-Boxes and Squares:
https://www.dropbox.com/s/c275u3236d4d44z/PatternV3_converted.gcode?dl=0

Seems there is no backlash, and if, its not that big like mentioned in all my other problem posts.

But thanks for the idea!

**Comment by thinkyhead** [comment]
@nophead That's a very good description of pulley backlash. If it's alright I'm going to borrow some of that verbiage and make a page for **Backlash** on the RepRap wiki, which somehow doesn't have one yet.

From what I understand the best test for backlash is to print circles at different sizes. If there's backlash it will be pretty obvious, because they will come out as ovals.

**Comment by nophead** [comment]
Yes fine to use my verbiage.

Circles reveal conventional backlash (where there is a dead spot during reversal) as you get two flats but they don't show pulley backlash as you are not going to notice an oval where one diameter is say 0.2mm less than the other.

The best way to show it is draw two lines that should meet end to end, but draw one having come from more than half a pulley circumference in one direction and the other the same distance from the other direction.

**Comment by AnHardt** [comment]
; Backlash test pattern

G21
G90
F500
G1 Z1
G1 X0 Y0
G1 X10 Y10
G1 Z0

G1 X20
G1 Y20
G1 X30
G1 Y10
G1 X40
G1 Y20
G1 X50
G1 Y10
G1 X60
G1 Y20

G1 X50
G1 Y30
G1 X40
G1 Y20
G1 X30
G1 Y30
G1 X20
G1 Y20
G1 X10
G1 Y30

G1 X20
G1 Y40
G1 X30
G1 Y30
G1 X40
G1 Y40
G1 X50
G1 Y30
G1 X60
G1 Y40

G1 X50
G1 Y50
G1 X40
G1 Y40
G1 X30
G1 Y50
G1 X20
G1 Y40
G1 X10
G1 Y50
...

![picture 134](https://cloud.githubusercontent.com/assets/211931/7565843/759d0ba6-f7f3-11e4-90a0-3d65d03f92c1.jpg)

Ideally there will be a perfect cross in the circle.

**Comment by VirToReal** [comment]
Good idea @AnHardt , my idea was not that efficient ;-)

I made that test and i guess its pritty good, sooo what next ;-)

![img_20150512_175227](https://cloud.githubusercontent.com/assets/10797066/7591885/9225208c-f8d0-11e4-84e3-946ab0b77c0e.jpg)

**Comment by AnHardt** [comment]
I'm seriously impressed. :-)
But without an idea for now. :-(

**Comment by VirToReal** [comment]
i tested with different arg segment and arc correction values (was not my idea :-) thanks to Holger :-), the idea was if there is a bug which will result in a mad behaviour. But it seems it quite works well. 
![dscn3136](https://cloud.githubusercontent.com/assets/10797066/7595542/fe532350-f8e7-11e4-9d7f-02e81392dd96.JPG)
![dscn3135](https://cloud.githubusercontent.com/assets/10797066/7595546/01fa2878-f8e8-11e4-828e-c4c0e4940fc9.JPG)

Will try it also with complete 0 values and will place feedback if something changes

**Comment by AnHardt** [comment]
As far i remember the ARC settings only apply to G2 and G3.
There are no arcs in your code.

**Comment by VirToReal** [comment]
this make sense, but take a look at the "V" letters, you will see differences on decreasing the values, it loses it smoothness :-O

**Comment by brainscan** [comment]
Was the paper completely flat when you took this picture? The lines aren't straight or parallel with each other. It looks good but when I placed a grid over your picture it shows they're not straight. It could be because I'm only working from the picture and it could distorted but it's worth checking again.

**Comment by AnHardt** [comment]
Anyone out there who knows a bit more about the ARC_DETECTION feature in Marlin?

**Comment by VirToReal** [comment]
@brainscan 
the picture was not completely flat (not much space in my cellar ;-), took the paper home and here we are:
![2015-05-12 22 45 34](https://cloud.githubusercontent.com/assets/10797066/7598339/e578a62c-f8f9-11e4-9efe-eb0b52250294.jpg)

**Comment by nophead** [comment]
AFAIK there is no arc detection in Marlin. That is done in some slicers. Marlin simply has G2 and G3 arc codes and I get the impression those are buggy, so nobody uses them for 3D printing.

**Comment by tonnico** [comment]
maybe a problem with the Dir-setup?
could you try this branch? https://github.com/Wurstnase/Marlin/tree/outbits_rework

**Comment by brainscan** [comment]
I did think that was probably the case but thought it worth checking. Still scratching my head on this one.

**Comment by VirToReal** [comment]
@Wurstnase 
what do you mean with Dir-Setup? 
~~~I tried your branch but got some issues on compiling:
https://www.dropbox.com/s/q6nwsnw7mspa2of/error_compile.txt?dl=0~~~

~~~it have to be something about the servo-engines ;-) (use one for auto-bed-leveling)
i rechecked the configuration files twice with a program called meld, everything should be fine, maybe you want so see them yourself:
https://www.dropbox.com/s/wvv3v1u8ah56aai/Configuration.h?dl=0
https://www.dropbox.com/s/lscrrsqoolv8nl6/Configuration_adv.h?dl=0~~~

EDIT:
sorry i'm stupid... i just moved your folder in my old one :-1:

**Comment by tonnico** [comment]
This is now in the development branch. You need only to download it again.

**Comment by VirToReal** [comment]
@Wurstnase 
i uploaded your branch but the results are sobering :-( see yourself:
![dscn3139](https://cloud.githubusercontent.com/assets/10797066/7732897/7d88a6b0-ff2b-11e4-9246-f6265ec9f538.JPG)
the new x-driver came today, i replaced the cheap one but the results (as mentioned) are the same. 

Now i figured something out...
It seems there is some kind of grid and with its resolution the g-code will be interpreted. I made some example:
Left: interpretation of the G-Code (made by pycam) with Chillipeppr. http://chilipeppr.com/tinyg
Right: the drawing of Inkscape (i poorly draw the letter O)
seems the gcode interpreter always makes this big steps when something changes on its grid. i think thats the cause of these issues but in fact i absolutely have no idea how this works anyway ;-)
(just a guess) 
![manipulation](https://cloud.githubusercontent.com/assets/10797066/7733137/86bad9f4-ff2d-11e4-9dff-b83b71a80fe4.png)
 i will test this out

**Comment by thinkyhead** [comment]
@HardRainbow The planner does approximate sometimes, and then there's the whole issue of acceleration on 4 axes… But, can you also show a detail of the same region of that "O" from your GCode, as it appears in a graphical GCode viewer, such as Repetier Host, Cura, Pronterface, etc?

**Comment by VirToReal** [comment]
The left picture was made with an gcode viewer 
 https://github.com/synthetos/TinyG/wiki/Chilipeppr
Same shape like the inkscape one

**Comment by thinkyhead** [comment]
@HardRainbow Great. Ok, well I suppose the next thing is to somehow log what's going on in the planner that might differ from the input.

**Comment by VirToReal** [comment]
I'm very very very sorry..... its an mechanical issue... 
It was very hard to find out, but its the Belt-Length on the Y-Axis. I tried out the Belt-Tightening at the very start. I was not able to see it because it only appears about 270° of the tightening-screw... 
My Skills in seeing necessary thinks wasnt good enough, too. I cant solve the problem by tightening the belt, seems it is quite to long... i looking for some software solutions..

I will post some pictures for showing the differences soon.

Thanks on all of you for helping me out! And sorry for all of this...

**Comment by brainscan** [comment]
I don't think there is any software that can correct mechanical problems. It would surely be more difficult than just fixing the mechanical problem and would be specific to one problem on one machine.

**Comment by VirToReal** [comment]
the only problem is, it cant be fixed mechanically ^^ the printer is just to big... only solution would be adapting to trapeze threaded rods. For the cost of speed... Also the layout have to change completely. I try to find out some other solution ;-) maybe there is one, if so, you will hear from me

**Comment by brainscan** [comment]
How about using pulleys and carbon fishing line like they do on some deltas and corexy, it can be put under huge tension.  I haven't tried it myself but from what I read about it would possibly suit a larger printer.

**Comment by VirToReal** [comment]
@brainscan Sounds nice but read this:

Sorry for this... its not over yet...
i was wondering because of the R Problem which shows no differences at all...first one is okay, the other one is wrong. So i made an test, with the R- Itself, and its last previous Z0 Point. I also moved that previous Z-Point on its hight. By moving above the starting point of the R-Surrounding, its **start** position is higher than bevore... 

Sorry, got no Cam here, will show pictures as soon as possible. Seems its some rounding mistake at all.. i tested with an friend of mine on his old 2013 Marlin firmware, its fine at that machine (most identical hardware setup). Would be nice if that G-Code file can be tested on other new Marlin Firmware. 

Here the Test-G-Code Files:
https://www.dropbox.com/s/m5eiy1ewvabkxik/VirToGravTest_R_Test3_converted.gcode?dl=0
https://www.dropbox.com/s/8atis8npzhtr5kp/VirToGravTest_O_Test2_converted.gcode?dl=0

Some Viewer to look at it:
http://chilipeppr.com/tinyg

Here the "R"-Test: (printed multiple times in same position)
![cimg0003](https://cloud.githubusercontent.com/assets/10797066/8145511/c8287514-120a-11e5-98fc-540ed6e1d7f3.JPG)

Here the "O"-Test: (printed multiple times in same position, row 1. got a different direction than row 2.)
![cimg0001](https://cloud.githubusercontent.com/assets/10797066/8145694/337f3242-1212-11e5-8b3a-95188bf90658.JPG)

Some Visual Help for the "R"-Test, First the Pen dropped left near the R-Surrounding, if the Pen dropps above the hight of the "R"-Surrounding the error appears (last 2. "R" letters)
![virtogravtest_r_test3_preview](https://cloud.githubusercontent.com/assets/10797066/8145681/0c68ffc6-1212-11e5-808f-926d51017e0b.png)

**Comment by VirToReal** [comment]
Ah sorry, it seems to be an mechanical problem too... i used the firmware we know it worked, it dont work on my machine so it must be mechanical. 

Sorry ... i close this and let it closed ;-) 

**_EDIT:**_
I got a solution:
If the Belt is tightened to strong, the gaps between the teeth’s of the belt have more space and the pulley wont fit well in it. If the belt is not tight enough, the belt itself begin so expand and relax in motion. The longer the belt, the more this problem will show.

Thanks to @AnHardt because of his Test-Pattern. I also made some picture-post in the previous post of that pattern. Just for notice: its not good enough if you want an perfect result. The lines have to fit perfectly to each other.
